source/common/network/tcp_listener_impl.cc

```diff
@@ -26,14 +26,38 @@ bool TcpListenerImpl::rejectCxOverGlobalLimit() const {
     return false;
   }
 
-  // If the connection limit is not set, don't limit the connections, but still track them.
-  // TODO(tonya11en): In integration tests, threadsafeSnapshot is necessary since the FakeUpstreams
-  // use a listener and do not run in a worker thread. In practice, this code path will always be
-  // run on a worker thread, but to prevent failed assertions in test environments, threadsafe
-  // snapshots must be used. This must be revisited.
-  const uint64_t global_cx_limit = runtime_.threadsafeSnapshot()->getInteger(
-      GlobalMaxCxRuntimeKey, std::numeric_limits<uint64_t>::max());
-  return AcceptedSocketImpl::acceptedSocketCount() >= global_cx_limit;
+  if (trackGlobalActiveCxLimitInOverloadManager()) {
+    // Check if deprecated runtime flag `overload.global_downstream_max_connections` is configured
+    // simultaneously with downstream connections monitor in overload manager.
+    if (runtime_.threadsafeSnapshot()->get(GlobalMaxCxRuntimeKey)) {
+      ENVOY_LOG_ONCE_MISC(
+          warn,
+          "Global downstream connections limits is configured via deprecated runtime key {} and in "
+          "{}. Using overload manager config.",
+          GlobalMaxCxRuntimeKey,
+          Server::OverloadProactiveResources::get().GlobalDownstreamMaxConnections);
+    }
+    // Try to allocate resource within overload manager. We do it once here, instead of checking if
+    // it is possible to allocate resource in this method and then actually allocating it later in
+    // the code to avoid race conditions.
+    return !(overload_state_->tryAllocateResource(
+        Server::OverloadProactiveResourceName::GlobalDownstreamMaxConnections, 1));
+  } else {
+    // If the connection limit is not set, don't limit the connections, but still track them.
+    // TODO(tonya11en): In integration tests, threadsafeSnapshot is necessary since the
+    // FakeUpstreams use a listener and do not run in a worker thread. In practice, this code path
+    // will always be run on a worker thread, but to prevent failed assertions in test environments,
+    // threadsafe snapshots must be used. This must be revisited.
+    const uint64_t global_cx_limit = runtime_.threadsafeSnapshot()->getInteger(
+        GlobalMaxCxRuntimeKey, std::numeric_limits<uint64_t>::max());
+    return AcceptedSocketImpl::acceptedSocketCount() >= global_cx_limit;
+  }
+}
+
+bool TcpListenerImpl::trackGlobalActiveCxLimitInOverloadManager() const {
+  return overload_state_ &&
```

Comment:
we'll compute this many times  (per accepted connection) over the lifetime of the TCP listener but we know at the ctor of the listener what the result will always be: as such I suggest just adding this as a bool to the object